pkg/controller/role/role_controller.go

```diff
@@ -227,48 +223,59 @@ func (r *ReconcileRole) Reconcile(request reconcile.Request) (reconcile.Result,
 			}
 		}
 
-		if template.Spec.Template.Spec.Containers[0].Image != sts.Spec.Template.Spec.Containers[0].Image {
-			reqLogger.Info("Updating container image")
-			sts.Spec.Template.Spec.Containers[0].Image = template.Spec.Template.Spec.Containers[0].Image
+		if !reflect.DeepEqual(template.Spec.Template.Spec, sts.Spec.Template.Spec) {
+			reqLogger.Info("Updating template spec")
+			sts.Spec.Template.Spec = template.Spec.Template.Spec
+			if err := r.client.Update(context.TODO(), &sts); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+
+		if template.Spec.UpdateStrategy.Type != sts.Spec.UpdateStrategy.Type {
+			reqLogger.Info("Updating sts UpdateStrategy", "new type", template.Spec.UpdateStrategy)
+			sts.Spec.UpdateStrategy = template.Spec.UpdateStrategy
 			if err := r.client.Update(context.TODO(), &sts); err != nil {
 				return reconcile.Result{}, err
 			}
 		}
 
-		sts.Spec.Template.Spec.Containers[0].Env = template.Spec.Template.Spec.Containers[0].Env
-		reqLogger.Info("Env variables", "vars", sts.Spec.Template.Spec.Containers[0].Env)
-		if err := r.client.Update(context.TODO(), &sts); err != nil {
-			return reconcile.Result{}, err
+		// immutable sts labels
+		for _, key := range []string{
+			"tarantool.io/replicaset-uuid",
+			"tarantool.io/vshardGroupName",
+		} {
+			template.ObjectMeta.Labels[key] = sts.ObjectMeta.Labels[key]
 		}
 
-		if templateRolesToAssign, ok := template.ObjectMeta.Annotations["tarantool.io/rolesToAssign"]; ok {
-			// check rolesToAssign from annotations
-			if templateRolesToAssign != sts.ObjectMeta.Annotations["tarantool.io/rolesToAssign"] {
-				reqLogger.Info("Updating replicaset rolesToAssign",
-					"from", sts.ObjectMeta.Annotations["tarantool.io/rolesToAssign"],
-					"to", templateRolesToAssign)
+		// immutable sts annotations
+		for _, key := range []string{
+			"tarantool.io/isBootstrapped",
+			"tarantool.io/replicaset-weight",
+		} {
+			template.ObjectMeta.Annotations[key] = sts.ObjectMeta.Annotations[key]
+		}
 
-				sts.ObjectMeta.Annotations["tarantool.io/rolesToAssign"] = templateRolesToAssign
-				sts.Spec.Template.Annotations["tarantool.io/rolesToAssign"] = templateRolesToAssign
+		// immutable container template labels
+		for _, key := range []string{
+			"tarantool.io/replicaset-uuid",
+			"tarantool.io/vshardGroupName",
+		} {
+			template.Spec.Template.ObjectMeta.Labels[key] = sts.Spec.Template.ObjectMeta.Labels[key]
```

Comment:
* no `r.client.Update` for this case
* expected update both Labels with Annotations